### PR TITLE
Prevent users to be created or updated with unknown Profiles

### PR DESCRIPTION
**Pull request by scottinet**
# Description

* Prevent `security:updateUser` to add unknown profiles to a user
* Fix wrong user ID provided to auth plugins during the rollback process when a user creation failed
* Refactor the `security:createUser` method to make it create the user in Kuzzle first, and then its credentials. This allows for more consistent behavior if an error occurs during the creation phase or even during the rollback one

# Solved issue

#893

### CODE REVIEW

File: lib/api/controllers/securityController.js [hunk 1]
```diff
@@ -449,7 +447,7 @@ class SecurityController {
     const pojoUser = request.input.body.content;
     pojoUser._id = request.input.resource._id || uuid.v4();
 
-    return Bluebird.coroutine(persistUserAndCredentials)(this.kuzzle, request, pojoUser);
+    return createGenerator(this.kuzzle, request, pojoUser);
```
Comment (kuzzle) [inline lib/api/controllers/securityController.js]: BLOCKER Move the declaration of "createGenerator" before this usage. [rule](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AVariableDeclarationAfterUsage)

File: lib/api/controllers/securityController.js [hunk 2]
```diff
@@ -420,8 +420,6 @@ class SecurityController {
   deleteUser(request) {
     assertHasId(request);
 
-    const deleteGenerator = Bluebird.coroutine(securityDeleteUserGenerator);
-
     return this.kuzzle.repositories.user.delete(request.input.resource._id, {refresh: request.input.args.refresh})
       .then(() => this.kuzzle.repositories.token.deleteByUserId(request.input.resource._id))
       .then(() => deleteGenerator(this.kuzzle, request))
```
Comment (kuzzle) [inline lib/api/controllers/securityController.js]: BLOCKER Move the declaration of "deleteGenerator" before this usage. [rule](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AVariableDeclarationAfterUsage)

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues
* ![BLOCKER][BLOCKER] 2 blocker

Watch the comments in this conversation to review them.

[BLOCKER]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER'

File: lib/api/controllers/securityController.js [hunk 3]
```diff
@@ -55,6 +55,144 @@ const
 
 /**
  * @param {Kuzzle} kuzzle
+ * @param {Request} request
+ * @param {object} pojoUser
+ */
+const createGenerator = Bluebird.coroutine(function* persistUserAndCredentials (kuzzle, request, pojoUser) {
```
Comment (benoitvidis) [inline lib/api/controllers/securityController.js]: nitpicking: `createGenerator` is not a generator,  `persistUserAndCredentials` is.

File: lib/api/controllers/securityController.js [hunk 4]
```diff
@@ -55,6 +55,144 @@ const
...
+  const loadedUser = yield Bluebird.resolve(kuzzle.repositories.user.load(pojoUser._id));
```
Comment (scottinet) [inline lib/api/controllers/securityController.js]: Didn't change that line of code, I'll look into it.

**Comment by scottinet** [comment]
@benoitvidis > fixed

File: lib/api/controllers/securityController.js [hunk 5]
```diff
@@ -55,6 +55,152 @@ const
...
+  // Throw in case of failure with the right KuzzleError object
+  const createdUser = yield kuzzle.repositories.user.hydrate(new User(), pojoUser)
+    .then(modifiedUser => kuzzle.repositories.user.persist(modifiedUser, {database: {method: 'create'}}))
+    .then(modifiedUser => formatProcessing.formatUserForSerialization(kuzzle, modifiedUser));
+
+  // Creating credentials
+  const errors = [];
+  let errorStep = null;
+
+  for (let i = 0; i < strategies.length; i++) {
```
Comment (ballinette) [inline lib/api/controllers/securityController.js]: why not `for (const strategy of strategies)` here ?

**Pull request merged**